subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1087 @@
+				spair->flags |= SPAIR_FLAG_NONBLOCK;
+			} else {
+				spair->flags &= ~SPAIR_FLAG_NONBLOCK;
+			}
+
+			res = 0;
+			goto out;
+		}
+
+		case ZFD_IOCTL_CLOSE: {
+			/* disconnect the remote endpoint */
+			spair->remote = -1;
+			spair_delete(spair);
+			res = 0;
+			goto out;
+		}
+
+		case ZFD_IOCTL_POLL_PREPARE: {
+			pfd = va_arg(args, struct zsock_pollfd *);
+			pev = va_arg(args, struct k_poll_event **);
+			pev_end = va_arg(args, struct k_poll_event *);
+
+			res = zsock_poll_prepare_ctx(obj, pfd, pev, pev_end);
+			goto out;
+		}
+
+		case ZFD_IOCTL_POLL_UPDATE: {
+			pfd = va_arg(args, struct zsock_pollfd *);
+			pev = va_arg(args, struct k_poll_event **);
+
+			res = zsock_poll_update_ctx(obj, pfd, pev);
+			goto out;
+		}
+
+		default: {
+			errno = EOPNOTSUPP;
+			res = -1;
+			goto out;
+		}
+	}
+
+out:
+	return res;
+}
+
+static int spair_bind(void *obj, const struct sockaddr *addr,
+		      socklen_t addrlen)
+{
+	(void)obj;
+	(void)addr;
+	(void)addrlen;
+
+	errno = EISCONN;
+	return -1;
+}
+
+static int spair_connect(void *obj, const struct sockaddr *addr,
+			 socklen_t addrlen)
+{
+	(void)obj;
+	(void)addr;
+	(void)addrlen;
+
+	errno = EISCONN;
+	return -1;
+}
+
+static int spair_listen(void *obj, int backlog)
+{
+	(void)obj;
+	(void)backlog;
+
+	errno = EINVAL;
+	return -1;
+}
+
+static int spair_accept(void *obj, struct sockaddr *addr,
+			socklen_t *addrlen)
+{
+	(void)obj;
+	(void)addr;
+	(void)addrlen;
+
+	errno = EOPNOTSUPP;
+	return -1;
+}
+
+static ssize_t spair_sendto(void *obj, const void *buf, size_t len,
+			    int flags, const struct sockaddr *dest_addr,
+				 socklen_t addrlen)
+{
+	(void)flags;
+	(void)dest_addr;
+	(void)addrlen;
+
+	return spair_write(obj, buf, len);
+}
+
+static ssize_t spair_sendmsg(void *obj, const struct msghdr *msg,
+			     int flags)
+{
+	(void)flags;
+
+	int res;
+
+	size_t len = 0;
+	struct spair *const spair = (struct spair *)obj;
+	const bool is_connected = sock_is_connected(spair);
+	const size_t avail = is_connected ? spair_write_avail(spair) : 0;
+	const bool is_nonblock = sock_is_nonblock(spair);
+
+	if (spair == NULL || msg == NULL) {
+		errno = EINVAL;
+		res = -1;
+		goto out;
+	}
+
+	for (size_t i = 0; i < msg->msg_iovlen; ++i) {
+		/* check & msg->msg_iov[i] */
+		/* check & msg->msg_iov[i].iov_base */
+		len += msg->msg_iov[i].iov_len;
+	}
+
+	if (!is_connected) {
+		errno = EPIPE;
+		res = -1;
+		goto out;
+	}
+
+	if (len == 0) {
+		res = 0;
+		goto out;
+	}
+
+	if (len > avail && is_nonblock) {
+		errno = EMSGSIZE;
+		res = -1;
+		goto out;
+	}
+
+	for (size_t i = 0; i < msg->msg_iovlen; ++i) {
+		res = spair_write(spair, msg->msg_iov[i].iov_base,
+				  msg->msg_iov[i].iov_len);
+		if (res == -1) {
+			goto out;
+		}
+	}
+
+	res = len;
+
+out:
+	return res;
+}
+
+static ssize_t spair_recvfrom(void *obj, void *buf, size_t max_len,
+			      int flags, struct sockaddr *src_addr,
+				   socklen_t *addrlen)
+{
+	(void)flags;
+	(void)src_addr;
```

Comment:
And like - not doing anything to src_addr/addrlen at all?

Reply:
If you would like to *suggest* that I do something with them, then please go ahead and *suggest* something.
Your review comment comes across as being somewhat condescending. Forgive me if that was not the intent.

Reply:
Did I miss something? Please be more specific.

Reply:
I'm trying to do first-level review for non-trivial code while multi-tasking on other things. If you'd prefer, I can put it off until I have a dedicated timeframe for it, but that won't be this week.

> Did I miss something? Please be more specific.

I don't know. A user asked us to return peer address, and passed buffer full of garbage and int*, pointing to size of that buffer. We're supposed to fill that buffer and put actual size into int*. We didn't do anything, so effectively returned garbage back to user. Is that what the POSIX wording says us to do for this type of sockets? I doubt that it, but can't jump to check per above, so suggest you to consider that case.

I don't know how to handle that case "right", there's a full spectrum from -ENOTSUPP (I would do it like that, given that recvfrom() is useful only for connectionless multi-agent transports like UDP) to actually write tests to see Linux behavior and implement the same.

Btw, I'd do -ENOTSUPP for sendmsg() either (unless I'd specifically need that), so extra points for implementing that. (The usual concern is how right it is, across socket types you support (sorry again, if you don't find that comment too cheerful, from my side, I just share the objective (ok, subjective) concerns)).

Reply:
I see your point. From
https://pubs.opengroup.org/onlinepubs/009695399/functions/recvfrom.html
it says

"If the address argument is not a null pointer and the protocol does not provide the source address of messages, the value stored in the object pointed to by address is unspecified."

It wasn't clear to me if the protocol (PF_UNIX) supports addressing on unnamed, connected sockets. for AF_UNIX, the name is the address. I'll throw a test together and run it on Linux.

If it is unspecified behaviour though, I would prefer to not modify any additional variables. If Linux gives an error, then I think it's fair to give an error as well.

Reply:
> It wasn't clear to me if the protocol (PF_UNIX) supports addressing on unnamed, connected sockets.

Ditto.

>  I'll throw a test together and run it on Linux.

I personally would find it acceptable for the initial version to return EOPNOTSUPP for recvfrom/sendto (unless you specifically need those calls right away, e.g. port a project which uses them).

Reply:
```cpp
#include <cerrno>
#include <iostream>
#include <stdexcept>

#include <sys/socket.h>
#include <sys/types.h>
#include <sys/un.h>
#include <unistd.h>

using namespace std;

int main(int argc, char *argv[]) {
        int r;
        int sv[2];
        char c;

        r = ::socketpair(AF_UNIX, SOCK_STREAM, 0, sv);
        if (-1 == r) {
                throw system_error(errno, system_category(), "socketpair(2)");
        }

        r = ::write(sv[0], "x", 1);
        if (-1 == r) {
                throw system_error(errno, system_category(), "write(2)");
        }

        ::sockaddr_un sa = {};
        ::socklen_t len = sizeof(sa);
        r = recvfrom(sv[1], &c, 1, 0, (::sockaddr *) &sa, &len);
        if (-1 == r) {
                throw system_error(errno, system_category(), "recvfrom(2)");
        }

        cout << "r: " << r << " len: " << len << endl;

        return 0;
```
```bash
$ ./foo 
r: 1 len: 0
```

Reply:
@pfalcon - does the above solution work for you (i.e. setting the addrlen to zero if it is not NULL)?

Reply:
> setting the addrlen to zero if it is not NULL

Yep, seems like a halfway between ENOTSUPP and "do it fully". Sounds good for initial version.